pkg/scheduler/core/generic_scheduler.go

```diff
@@ -104,15 +104,19 @@ const (
 // Error returns detailed information of why the pod failed to fit on each node
 func (f *FitError) Error() string {
 	reasons := make(map[string]int)
-	for _, predicates := range f.FailedPredicates {
-		for _, pred := range predicates {
-			reasons[pred.GetReason()]++
+	for _, statuses := range f.FailedFilters {
+		for _, status := range statuses {
+			reasons[status.Message()]++
 		}
 	}
 
 	sortReasonsHistogram := func() []string {
 		reasonStrings := []string{}
 		for k, v := range reasons {
+			predicateFailureError, ok := predicates.FindPredicateFailureError(k)
+			if ok {
+				k = predicateFailureError.GetReason()
+			}
```

Comment:
why are we changing the key that gets printed on line 120 from the predicate name to the description? I think it should stay as it was before, we are trying to build a histogram of the how many nodes were rejected by each predicate.

Reply:
Sorry, I don't get it. I think we are using the predicate description `pred.GetReason()` to build the message before. Could you explain a little bit more of this?

https://github.com/kubernetes/kubernetes/pull/78097/files#diff-c237cdd9e4cb201118ca380732d7f361L107-L111

Reply:
It is indeed confusing, it took me a while to trace it, pred.GetReason() actually returns the predicate name, it is an instance of [FailureReason](https://github.com/kubernetes/kubernetes/blob/f101466d2e4d96854c80f58203de2cc6b5aaeb6a/pkg/scheduler/algorithm/predicates/error.go#L153) which implements [PredicateFailureReason interface](https://github.com/kubernetes/kubernetes/blob/f101466d2e4d96854c80f58203de2cc6b5aaeb6a/pkg/scheduler/algorithm/predicates/error.go#L138:6) this is not an instance of [PredicateFailureError](https://github.com/kubernetes/kubernetes/blob/f101466d2e4d96854c80f58203de2cc6b5aaeb6a/pkg/scheduler/algorithm/predicates/error.go#L133).

Reply:
> It is indeed confusing, it took me a while to trace it, pred.GetReason() actually returns the predicate name, it is an instance of [FailureReason](https://github.com/kubernetes/kubernetes/blob/f101466d2e4d96854c80f58203de2cc6b5aaeb6a/pkg/scheduler/algorithm/predicates/error.go#L153) which implements [PredicateFailureReason interface](https://github.com/kubernetes/kubernetes/blob/f101466d2e4d96854c80f58203de2cc6b5aaeb6a/pkg/scheduler/algorithm/predicates/error.go#L138:6), this is not an instance of [PredicateFailureError](https://github.com/kubernetes/kubernetes/blob/f101466d2e4d96854c80f58203de2cc6b5aaeb6a/pkg/scheduler/algorithm/predicates/error.go#L133).

I searched the usage of `FailureReason`, and it is only used in the [extender](https://github.com/kubernetes/kubernetes/blob/7f9dd949e01189709b7ec7955204e04378aa64cb/pkg/scheduler/core/generic_scheduler.go#L558-L563
) and [filter plugins](https://github.com/kubernetes/kubernetes/blob/7f9dd949e01189709b7ec7955204e04378aa64cb/pkg/scheduler/core/generic_scheduler.go#L507-L515), which I think it was created to help bulid a message in `FitError.Error`

The `failedPredicates` returned from the predicates functions are all `PredicateFailureError`.

I'll open a follow-up PR to remove this from the codebase since this is no longer useful after this get merged.

Reply:
Wow, this is very confusing and indeed needs a good cleanup. I still don't understand why we need this lookup if we [already set the framework.Status message](https://github.com/kubernetes/kubernetes/pull/78097/files#diff-c237cdd9e4cb201118ca380732d7f361R676) to the description returned by GetReason?

Reply:
I understand now what you are trying to do, unresolvablePredicateExists needs a key to lookup in the unresolvablePredicateFailureErrors map, and since now we changed the type, that key longer can be the pointer to one of the predefined PredicateFailureError structs, and so Bobby suggested to use the predicate name.

The main confusing part to me was that [here](https://github.com/kubernetes/kubernetes/pull/78097/files#diff-c237cdd9e4cb201118ca380732d7f361R676) we use GetReason instead of getting the name, which we actually don't have a method to retrieve in PredicateFailureReason interface.

In all cases, why are we trying to cleanup error handling in this PR? is this related to implementing the post-filter extension point in any way?

Reply:
> In all cases, why are we trying to cleanup error handling in this PR? is this related to implementing the post-filter extension point in any way?

Because we should return `Status` instead of `FailurePredicate` as Boddy suggested [here](https://github.com/kubernetes/kubernetes/pull/78097#discussion_r307072431)

Reply:
> > In all cases, why are we trying to cleanup error handling in this PR? is this related to implementing the post-filter extension point in any way?
> 
> Because we should return `Status` instead of `FailurePredicate` as Boddy suggested [here](https://github.com/kubernetes/kubernetes/pull/78097#discussion_r307072431)

The comment is referring to an outdated code, and your are amending you commits, so I can't really assess properly, but I suspect that also has nothing to do with post-filter, right?

Reply:
`findNodesThatFit` needs to return a `FailedFilterMap` which is related to the post-filter

Reply:
hmm, I don't agree with this dependency, converting FailurePredicate to Status looks like a hack to me, I will comment on the other comment.

But if we are to keep this logic, I guess you need to add a method to PredicateFailureReason to get the predicate name.